### ISSUE TITLE
Differentiable Optimizers

### ISSUE THREAD

**Issue by vincentqb** [pytorch/pytorch#39279]
## 🚀 Feature

Let's design below what an API for differential optimizers would look like. The goal is to facilitate the implementation of differential optimizer for libraries such as [higher](https://github.com/facebookresearch/higher/). 

Related discussions in #12659 and #32005, and [Internal document](https://fb.quip.com/7fsdAavbPGJ0)

cc @albanD @egrefen

**Comment by vincentqb** [comment]
Let's look at a conceptual example.

```python
# Let's assume these are both subclasses of torch.nn.Module, but that they 
# are third-party code I don't want to touch, and that I might be loading 
# saved weights into them
model = MyModel(input_size=input_size, output_size=num_classes)
metamodel = MyModel(input_size=input_size, output_size=1)

step = 0

opt = MyOptim(model.parameters())
metaopt = MyOptim(metamodel.parameters())

for epoch in num_epochs:

    # Normal training loop
    for (images, labels) in my_dataset:
        opt.zero_grad()
        step += 1

        # training step
        preds = model(images)
        loss_weight = metamodel(images)
        loss = my_loss_function(preds, labels) * loss_weight

        # update model params
        loss.backward()
        opt.step()

        print("Training loss:", loss.item(), "at step", step)

        # Every meta_update_freq steps (e.g. once per epoch)...
        if step % meta_update_freq == 0:

            # ... update the parameters of the model num_meta_updates times
            for meta_step in num_meta_updates:
                metaopt.zero_grad()

                inner_model = model.clone()  # safely branch off model state
                inner_opt = opt.clone()  # safely branch of optimizer state

                inner_opt.differentiable = True  # set some state to say the optimizer operates differentiably

                # Unroll the training loop num_inner_steps training steps
                for inner_step in num_inner_steps:
                    inner_opt.zero_grad()  # doesn't delete information required to do "backprop-through-backprop"
                    images, labels = my_dataset.next()
                    preds = inner_model(images)
                    loss_weight = metamodel(images)
                    loss = my_loss_function(preds, labels) * loss_weight
                    loss.backward()
                    inner_opt.step()  # updates params but keeps backward graph to beginning of unrolled loop

                # Compute validation loss on the latest version of the unrolled model
                val_images, val_labels = my_validation_set.next()
                val_preds = inner_model(val_images)
                meta_loss = my_loss_function(val_preds, val_labels)  # no loss_weight for meta-loss
                

                # backprop to parameters of metamodel and update them
                meta_loss.backward()  # this should propagate to every use of the parameters within the unrolled loop
                metaopt.step()

                print("Meta loss:", meta_loss.item(), "at meta_step", meta_step, "of step", step)

    # Run a normal validation loop
    for (val_images, val_labels) in my_validation_set:
        val_preds = model(val_images)
        val_loss = my_loss_function(val_preds, val_labels)  # report unweighted validation loss
        print("Validation loss:", val_loss.item(), "at step", step)
```

**Comment by vincentqb** [comment]
It seems like the API we want would look more like the following. Thoughts?

```python
inner_model = nnModuleModel(input_size=input_size, output_size=num_classes)
metamodel = nnModuleModel(input_size=input_size, output_size=1)

step = 0

metaopt = Optim(metamodel.parameters(), with_no_grad=True)  # conventional optimizer with "with no_grad"

for epoch in num_epochs:

    # Normal training loop
    for (images, labels) in my_dataset:
        opt.zero_grad()
        step += 1

        # training step
        preds = model(images)
        loss_weight = metamodel(images)
        loss = my_loss_function(preds, labels) * loss_weight

        # update model params
        loss.backward()
        opt.step()

        print("Training loss:", loss.item(), "at step", step)

        # Every meta_update_freq steps (e.g. once per epoch)...
        if step % meta_update_freq == 0:

            state_dict = inner_model.state_dict()

            # ... update the parameters of the model num_meta_updates times
            for meta_step in num_meta_updates:
                metaopt.zero_grad()

                inner_model.load_state_dict(state_dict)  # not currently differentiable
                inner_opt = Optim(model.parameters(), with_no_grad=False)  # optimizer without "with no_grad"

                # Unroll the training loop num_inner_steps training steps
                for inner_step in num_inner_steps:
                    inner_opt.zero_grad()  # doesn't delete information required to do "backprop-through-backprop"
                    images, labels = my_dataset.next()
                    preds = inner_model(images)
                    loss_weight = metamodel(images)
                    loss = my_loss_function(preds, labels) * loss_weight
                    loss.backward(to=inner_model)  # exposes autograd.grad(loss, inner_model) to keep backward graph to beginning of unrolled loop but take the derivative only within this inner_step
                    inner_opt.step()

                # Compute validation loss on the latest version of the unrolled model
                val_images, val_labels = my_validation_set.next()
                val_preds = inner_model(val_images)
                meta_loss = my_loss_function(val_preds, val_labels)  # no loss_weight for meta-loss

                # backprop to parameters of metamodel and update them
                meta_loss.backward()  # this should propagate to every use of the parameters within the unrolled loop
                metaopt.step()

                print("Meta loss:", meta_loss.item(), "at meta_step", meta_step, "of step", step)

    # Run a normal validation loop
    for (val_images, val_labels) in my_validation_set:
        val_preds = model(val_images)
        val_loss = my_loss_function(val_preds, val_labels)  # report unweighted validation loss
        print("Validation loss:", val_loss.item(), "at step", step)
```

To do this, I see
- [ ] An addition to the backward interface to provide an easier access to `autograd.grad(loss, previous_model)`
- [ ] A flag in the optimizer constructor to remove the torch.no_grad from the optimizer.
- [x] Cleaning up the .data #34211

**Comment by jspisak** [comment]
cc @tristandeleu and seba-1511 for comments and inputs. Please take a look!

**Comment by seba-1511** [comment]
Thanks @jspisak, those are exciting suggestions!

`model.clone()` and `opt.clone()` are very PyTorch-y and keep a clear separation between the original and the cloned module. This is helpful when you need both modules, e.g. for importance weights in meta-rl. A differentiable `load_state_dict()` also works, but is it less convenient? (Related: [clone_module()](https://github.com/learnables/learn2learn/blob/master/learn2learn/utils.py#L51) in learn2learn.)

`loss.backward(to=model)` is also useful; but why depart from the `torch.autograd.grad()` API? Could we have `loss.backward(inputs=model.parameters())` so I can pass any list of variables to `inputs`, not just a Module?

I don't have a strong opinion on differentiable optimizers.

**Comment by vincentqb** [comment]
Following offline discussion with @seba-1511. An alternative to [comment](https://github.com/pytorch/pytorch/issues/39279#issuecomment-636238638),
- [ ] Instead of a differentiable `load_state_dict`, a new functional optimizer would (1) avoid creating a new optimizer instance in the inner loop, and (2) allow sharing parameters across tasks (e.g. [Reptile](https://openai.com/blog/reptile/)).
- [ ] As mentioned before, an addition to the backward interface to provide an easier access to `autograd.grad(loss, previous_model)`
- [ ] Instead of `with_no_grad=False`, we could have `differentiable=True`. This would also update the model in place by extending the computational graph without breaking it, e.g. for SGD:
```python
for name, update in zip(model._parameters, updates):
    model._parameters[name] = model._parameters[name] - lr * update
```
instead what it currently does:
```python
for p, update in zip(model.parameters(), updates):
    p.add_(-lr, update)
```
Thoughts?

**Comment by vincentqb** [comment]
> * [ ]  Instead of a differentiable `load_state_dict`, a new functional optimizer would (1) avoid creating a new optimizer instance in the inner loop, and (2) allow sharing parameters across tasks (e.g. [Reptile](https://openai.com/blog/reptile/)). (see also [internal doc](https://fb.quip.com/XMLfA0XR4tnp))

For those interested in functional optimizers in this context, I would love to hear if #44715 is aligned with how you were envisioning them :)

**Comment by egrefen** [comment]
Thanks for the code snippet, @vincentqb, and sorry for taking so long to get back to you. Just as a point of clarification, would the idea be that at the end of the meta-training loop, the `inner_module` state would correspond to its state at the end of the latest inner loop, or would it revert to how it was before the meta-training loop was done? If the latter, should we expect a `inner_model.load_state_dict(state_dict)` at the end of meta-training?

**Comment by egrefen** [comment]
I agree @seba-1511 that letting the `to=` keyword in `backward` take an iterable of parameters/leaf-nodes in the backward graph offers more flexibility (even outside of the meta-learning case).

**Comment by albanD** [comment]
@egrefen and @seba-1511 for the autograd API, the inputs= choice is the one that was made indeed and this is already available in master: https://pytorch.org/docs/master/autograd.html#torch.autograd.backward

Note that, for now, the given inputs must be leafs. We can lift that constraint in the future if needed.

**Comment by egrefen** [comment]
Goes to show my knowledge is out of date. That's super!
